underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java

```diff
@@ -0,0 +1,406 @@
+   * @param s3Client the Amazon S3 client to upload the file with
+   * @param executor a thread pool executor
+   */
+  public S3ALowLevelOutputStream(String bucketName, String key, AmazonS3 s3Client,
+      ExecutorService executor) {
+    Preconditions.checkArgument(bucketName != null && !bucketName.isEmpty(), "Bucket name must "
+        + "not be null or empty.");
+    mBucketName = bucketName;
+    mClient = s3Client;
+    mExecutor = MoreExecutors.listeningDecorator(executor);
+    try {
+      mHash = MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      LOG.warn("Algorithm not available for MD5 hash.", e);
+      mHash = null;
+    }
+    mKey = key;
+    // Partition size should be at least 5 MB, since S3 low-level multipart upload does not
+    // accept intermediate part smaller than 5 MB.
+    long partSize = Configuration.getBytes(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_PARTITION_SIZE);
+    mPartitionSize = partSize < 5 * Constants.MB ? 5 * Constants.MB : partSize;
+    mPartNumber = new AtomicInteger(1);
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    mSingleCharWrite[0] = (byte) b;
+    write(mSingleCharWrite, 0, 1);
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (mUploadId == null) {
+      initMultiPartUpload();
+    }
+    if (b == null || len == 0) {
+      return;
+    }
+    if (mFile == null) {
+      initNewFile();
+    }
+    if (mOffset + len < mPartitionSize) {
+      mLocalOutputStream.write(b, off, len);
+      mOffset += len;
+    } else {
+      int firstLen = (int) (mPartitionSize - mOffset);
+      mLocalOutputStream.write(b, off, firstLen);
+      mOffset += firstLen;
+      uploadPart();
+      write(b, off + firstLen, len - firstLen);
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    // We try to minimize the time use to close()
+    // because Fuse release() method which calls close() is async.
+    // In flush(), we upload the current writing file if it is bigger than 5 MB,
+    // and wait for all current upload to complete.
+    mLocalOutputStream.flush();
+    if (mOffset > UPLOAD_THRESHOLD) {
+      uploadPart();
+    }
+    waitForAllPartsUpload();
+    mFutureTags = new ArrayList<>();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    try {
+      if (mFile != null) {
+        mLocalOutputStream.close();
+        int partNumber = mPartNumber.getAndIncrement();
+        final UploadPartRequest uploadRequest = new UploadPartRequest()
+            .withBucketName(mBucketName)
+            .withKey(mKey)
+            .withUploadId(mUploadId)
+            .withPartNumber(partNumber)
+            .withFile(mFile)
+            .withPartSize(mFile.length());
+        uploadRequest.setLastPart(true);
```

Comment:
if each time when write is called, length is exactly mPartitionSize, then when `close` is called, `mFile` is `null`, so `setLastPart(true)` will never be called, does `setLastPart` must be called for the last part?